Add invite endpoint v1 and v2

agraven opened the pull request:
Still needs v2.

I was a bit unsure about what names to give the helper structs, and also what types to use for timestamps

agraven committed:
Add invite endpoint v1

jplatte commented on ruma-federation-api/src/membership/create_invite/v1.rs:
```diff
@@ -0,0 +1,157 @@
+use js_int::UInt;
+use ruma_api::ruma_api;
+use ruma_events::room::member::MemberEventContent;
+use ruma_events::EventType;
```
Please merge these imports.

jplatte commented on ruma-federation-api/src/membership/create_invite/v1.rs:
```diff
@@ -0,0 +1,157 @@
+        room_id: RoomId,
+
+        /// The event ID for the invite event, generated by the inviting server.
+        #[serde(rename = "eventId")]
```
I don't think serde attributes work on `#[ruma_api(path)]` fields. This is probably a leftover?

agraven replied:
No this was my forgetting that you get to set the parameter names for paths, and that you don't have to name them exactly as the spec does

jplatte commented on ruma-federation-api/src/membership/create_invite/v1.rs:
```diff
@@ -0,0 +1,157 @@
+
+impl Request {
+    /// Creates a new `Request` with the given parameters.
+    pub fn new(
```
For a type with this many fields, let's create an [init struct](https://github.com/ruma/ruma/blob/1fbf05d71ffa94f0571ca9850aa5c112ddd85dff/ruma-common/src/push.rs#L77).

jplatte commented on ruma-federation-api/src/membership/create_invite/v1.rs:
```diff
@@ -0,0 +1,157 @@
+
+impl Response {
+    /// Creates a new `Response` with the given parameters
+    pub fn new(
```
Same here.

agraven committed:
Add init structs, remove unneeded attributes, import fixes

agraven committed:
Add invite endpoint v2

jplatte commented on ruma-federation-api/src/membership/create_invite/v1.rs:
```diff
@@ -0,0 +1,146 @@
+    content: MemberEventContent,
+}
+
+impl Request {
```
Can you make this a `From` conversion instead?

```suggestion
impl From<RequestInit> for Request {
```

jplatte commented on ruma-federation-api/src/membership/create_invite/v1.rs:
```diff
@@ -0,0 +1,146 @@
+    }
+}
+
+impl Response {
```
Same here.

jplatte commented on ruma-federation-api/src/membership/create_invite/mod.rs:
```diff
@@ -0,0 +1,50 @@
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct StrippedState {
+    /// The `content` for the event.
+    content: Value,
```
All fields are currently private, they should be `pub`lic.

jplatte commented on ruma-federation-api/src/membership/create_invite/mod.rs:
```diff
@@ -0,0 +1,50 @@
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct InviteEvent {
+    /// The matrix ID of the user who sent the original `m.room.third_party_invite`.
+    sender: UserId,
```
Same here.

jplatte commented on ruma-federation-api/src/membership/create_invite/v1.rs:
```diff
@@ -0,0 +1,146 @@
+    request: {
+        /// The room ID that the user is being invited to.
+        #[ruma_api(path)]
+        room_id: RoomId,
```
Same everywhere in this module.

jplatte commented on ruma-federation-api/src/membership/create_invite/v2.rs:
```diff
@@ -0,0 +1,90 @@
+    }
+}
+
+impl Response {
```
`Response` should have a regular constructor since there is only one field to initialize. This code can still be useful though as `impl From<InviteEventInit> for InviteEvent` – `InviteEvent` should also be `#[non_exhaustive]` as well as the request and response blocks, these accept attributes now; see also the new section of CONTRIBUTING.md: [(Type) Privacy and Forwards Compatibility](https://github.com/ruma/ruma/blob/master/CONTRIBUTING.md#type-privacy-and-forwards-compatiblity).

agraven committed:
Corrections for structs

Make members public where needed, add non-exhaustive attributes, correct
constructors

q-b commented:
Shouldn't `ruma-federation-api/src/membership/create_invite/mod.rs` be renamed  `ruma-federation-api/src/membership/create_invite.rs`?

jplatte commented:
Yes, though until recently we had mod.rs files all over ruma-appservice-api and ruma-federation-api for some reason. Thanks for noticing!

agraven commented:
Since it was the pattern all other endpoints were following, I assumed it was standard practice to use mod.rs specifically for modules containing `v1`, `v2`, etc, I didn't realize this wasn't supposed to be the case. Thanks for clearing things up!

agraven committed:
Don't use mod.rs naming

jplatte commented on ruma-federation-api/src/membership/create_invite.rs:
```diff
@@ -0,0 +1,83 @@
+use ruma_events::{room::member::MemberEventContent, EventType};
+use ruma_identifiers::{ServerName, UserId};
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
```
Please import-rename for consistency as noted in `CONTRIBUTING.md`.

```suggestion
use serde_json::Value as JsonValue;
```

jplatte commented on ruma-federation-api/src/membership/create_invite.rs:
```diff
@@ -0,0 +1,83 @@
+
+/// The invite event sent as a response.
+#[non_exhaustive]
+#[derive(Clone, Debug, Serialize, Deserialize)]
```
Derives should come forst, for consistency.

jplatte commented on ruma-federation-api/src/membership/create_invite/v1.rs:
```diff
@@ -0,0 +1,121 @@
+        /// The response invite event
+        #[ruma_api(body)]
+        #[serde(with = "crate::serde::invite_response")]
+        invite_response: InviteEvent,
```
Maybe just name this `event` too? Since it's `#[ruma_api(body)]` its name doesn't matter, but `response.invite_response` doesn't seem so nice.

agraven replied:
That does seem a lot nicer, I forgot the attribute means the name doesn't matter

agraven committed:
Add changelog entry, corrections

Correct import, attribute, struct member name
